File: src/server/natsProxy.ts
import {
  connect,
  NatsConnection,
  Subscription,
  StringCodec,
  NatsError,
  MsgHdrs,
} from "nats";
import { WebSocket } from "ws";
import { randomUUID } from "crypto";

const sc = StringCodec();

function headersToRecord(hdrs: MsgHdrs | undefined): Record<string, string> {
  if (!hdrs) return {};
  const result: Record<string, string> = {};
  for (const key of hdrs.keys()) {
    result[key] = hdrs.get(key) ?? "";
  }
  return result;
}

function send(ws: WebSocket, data: object) {
  if (ws.readyState === WebSocket.OPEN) {
    ws.send(JSON.stringify(data));
  }
}

export class NatsProxy {
  private nc: NatsConnection | null = null;
  private discoverySub: Subscription | null = null;
  private userSubs = new Map<string, Subscription>();
  private seenSubjects = new Set<string>();
  private ws: WebSocket;

  constructor(ws: WebSocket) {
    this.ws = ws;
  }

  async connectToServer(url: string) {
    await this.disconnect();
    this.seenSubjects.clear();

    try {
      this.nc = await connect({ servers: url });
      send(this.ws, { type: "connected", serverInfo: this.nc.info });
      this.startDiscovery();
      this.watchClosed();
    } catch (err) {
      send(this.ws, {
        type: "error",
        message: `Failed to connect to ${url}: ${(err as Error).message}`,
      });
    }
  }

  private watchClosed() {
    (async () => {
      if (!this.nc) return;
      for await (const s of this.nc.status()) {
        if (s.type === "disconnect" || s.type === "error") {
          send(this.ws, { type: "error", message: `NATS: ${s.type}` });
        }
      }
    })().catch(() => {});
  }

  private startDiscovery() {
    if (!this.nc) return;
    this.discoverySub = this.nc.subscribe(">", {
      callback: (err, msg) => {
        if (err) return;

        const subject = msg.subject;
        const payloadBytes = msg.data;
        const bufData = Buffer.from(payloadBytes);
        const asUtf8 = bufData.toString("utf8");
        const isText = Buffer.from(asUtf8, "utf8").equals(bufData);
        const payload = isText ? asUtf8 : bufData.toString("base64");
        const encoding = isText ? "text" : "base64";

        if (!this.seenSubjects.has(subject)) {
          this.seenSubjects.add(subject);
          send(this.ws, { type: "subjects", subjects: [subject] });
        }

        send(this.ws, {
          type: "message",
          id: randomUUID(),
          subject,
          replyTo: msg.reply ?? null,
          payload,
          encoding,
          headers: headersToRecord(msg.headers),
          timestamp: Date.now(),
          size: payloadBytes.length,
        });
      },
    });
  }

  async publish(subject: string, payload: string, headers?: Record<string, string>) {
    if (!this.nc) {
      send(this.ws, { type: "error", message: "Not connected" });
      return;
    }
    try {
      const opts: { headers?: MsgHdrs } = {};
      if (headers && Object.keys(headers).length > 0) {
        const { headers: createHeaders } = await import("nats");
        const h = createHeaders();
        for (const [k, v] of Object.entries(headers)) {
          h.append(k, v);
        }
        opts.headers = h;
      }
      this.nc.publish(subject, sc.encode(payload), opts);
      send(this.ws, { type: "published", subject });
    } catch (err) {
      send(this.ws, {
        type: "error",
        message: `Publish failed: ${(err as Error).message}`,
      });
    }
  }

  async disconnect() {
    this.discoverySub?.unsubscribe();
    this.discoverySub = null;
    for (const sub of this.userSubs.values()) sub.unsubscribe();
    this.userSubs.clear();
    if (this.nc) {
      await this.nc.drain().catch(() => {});
      this.nc = null;
    }
  }
}

File: server_py/nats_proxy.py
import asyncio
import base64
import uuid
import time
from typing import TYPE_CHECKING

import nats

if TYPE_CHECKING:
    from aiohttp.web import WebSocketResponse


class NatsProxy:
    def __init__(self, ws: "WebSocketResponse") -> None:
        self._ws = ws
        self._nc: nats.aio.client.Client | None = None
        self._discovery_sub = None
        self._seen_subjects: set[str] = set()

    async def connect(self, url: str) -> None:
        await self.disconnect()
        self._seen_subjects.clear()

        async def error_cb(e: Exception) -> None:
            await self._send({"type": "error", "message": str(e)})

        async def closed_cb() -> None:
            await self._send({"type": "error", "message": "NATS connection closed"})

        try:
            self._nc = await nats.connect(
                url,
                error_cb=error_cb,
                closed_cb=closed_cb,
            )
            info = getattr(self._nc, "_server_info", {}) or {}
            await self._send({"type": "connected", "serverInfo": info})
            await self._start_discovery()
        except Exception as e:
            await self._send({"type": "error", "message": f"Failed to connect to {url}: {e}"})

    async def publish(self, subject: str, payload: str, headers: dict[str, str] | None = None) -> None:
        if self._nc is None:
            await self._send({"type": "error", "message": "Not connected"})
            return
        try:
            data = payload.encode("utf-8")
            await self._nc.publish(subject, data, headers=headers or None)
            await self._send({"type": "published", "subject": subject})
        except Exception as e:
            await self._send({"type": "error", "message": f"Publish failed: {e}"})

    async def disconnect(self) -> None:
        if self._discovery_sub is not None:
            try:
                await self._discovery_sub.unsubscribe()
            except Exception:
                pass
            self._discovery_sub = None
        if self._nc is not None:
            try:
                await self._nc.drain()
            except Exception:
                pass
            self._nc = None

    async def _start_discovery(self) -> None:
        if self._nc is None:
            return

        async def on_msg(msg) -> None:
            data: bytes = msg.data
            try:
                text = data.decode("utf-8")
                if text.encode("utf-8") == data:
                    payload, encoding = text, "text"
                else:
                    raise ValueError
            except (UnicodeDecodeError, ValueError):
                payload = base64.b64encode(data).decode("ascii")
                encoding = "base64"

            raw_headers: dict | None = msg.headers
            headers: dict[str, str] = {}
            if raw_headers:
                for k, v in raw_headers.items():
                    headers[k] = v[0] if isinstance(v, list) else str(v)

            subject = msg.subject
            if subject not in self._seen_subjects:
                self._seen_subjects.add(subject)
                await self._send({"type": "subjects", "subjects": [subject]})

            await self._send({
                "type": "message",
                "id": str(uuid.uuid4()),
                "subject": subject,
                "replyTo": msg.reply or None,
                "payload": payload,
                "encoding": encoding,
                "headers": headers,
                "timestamp": int(time.time() * 1000),
                "size": len(data),
            })

        self._discovery_sub = await self._nc.subscribe(">", cb=on_msg)

    async def _send(self, data: dict) -> None:
        try:
            if not self._ws.closed:
                await self._ws.send_json(data)
        except Exception:
            pass
